packages/contracts/contracts/SmartInvoiceSplitEscrow.sol

```diff
@@ -0,0 +1,587 @@
+    }
+
+    /**
+     * @dev Verifies the client and contract are paired
+     */
+    function verify() external {
+        require(msg.sender == client, "!client");
+        emit Verified(client, address(this));
+    }
+
+    /**
+     * @dev Adds milestones without extra details.
+     * @param _milestones The array of new milestones to be added
+     */
+    function addMilestones(uint256[] calldata _milestones) external {
+        _addMilestones(_milestones, bytes32(0));
+    }
+
+    /**
+     * @dev Adds milestones with extra details.
+     * @param _milestones The array of new milestones to be added
+     * @param _details Additional details for the milestones
+     */
+    function addMilestones(uint256[] calldata _milestones, bytes32 _details)
+        external
+    {
+        _addMilestones(_milestones, _details);
+    }
+
+    /**
+     * @dev Internal function to add milestones and update the contract state.
+     * @param _milestones The array of new milestones to be added
+     * @param _details Additional details for the milestones
+     */
+    function _addMilestones(uint256[] calldata _milestones, bytes32 _details)
+        internal
+    {
+        require(!locked, "locked");
+        require(block.timestamp < terminationTime, "terminated");
+        require(_msgSender() == client || _msgSender() == provider, "!party");
+        require(_milestones.length > 0, "no milestones are being added");
+        require(_milestones.length <= 10, "only 10 new milestones at a time");
+
+        uint256 newLength = amounts.length + _milestones.length;
+        uint256[] memory baseArray = new uint256[](newLength);
+        uint256 newTotal = total;
+
+        for (uint256 i = 0; i < amounts.length; i++) {
+            baseArray[i] = amounts[i];
+        }
+        for (uint256 i = amounts.length; i < newLength; i++) {
+            baseArray[i] = _milestones[i - amounts.length];
+            newTotal += _milestones[i - amounts.length];
+        }
+
+        total = newTotal;
+        amounts = baseArray;
+
+        if (_details != bytes32(0)) {
+            details = _details;
+            emit DetailsUpdated(msg.sender, _details);
+        }
+
+        emit MilestonesAdded(msg.sender, address(this), _milestones);
+    }
+
+    /**
+     * @dev Returns the amounts associated with the milestones.
+     * @return An array of amounts for each milestone
+     */
+    function getAmounts() public view returns (uint256[] memory) {
+        return amounts;
+    }
+
+    /**
+     * @dev Internal function to release funds from the contract to the provider.
+     */
+    function _release(uint256 _milestone) internal {
+        // client transfers locker milestone funds to provider and dao
+
+        require(!locked, "locked");
+        require(_msgSender() == client, "!client");
+        require(_milestone < amounts.length, "invalid milestone");
+
+        uint256 balance = IERC20(token).balanceOf(address(this));
+
+        if (_remaining >= 1) {
```

Comment:
FYI performing the else conditional (`_remaining < 1`) first will save a bit of gas by using 1 less operation.